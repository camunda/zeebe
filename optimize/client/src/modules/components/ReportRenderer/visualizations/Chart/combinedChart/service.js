/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';

const {formatReportResult} = formatters;

// Override the default generate legend's labels function
// This is done to modify the colors retrieval method of the legend squares and filter unneeded labels
export function generateLegendLabels(chart) {
  const data = chart.data;
  return data.datasets.length
    ? data.datasets
        .map(function (dataset) {
          return {
            text: dataset.label,
            fillStyle: !dataset.backgroundColor.length
              ? dataset.backgroundColor
              : dataset.legendColor,
            strokeStyle: dataset.legendColor,
          };
        }, this)
        .filter((dataset) => {
          return dataset.text;
        })
    : [];
}

export function getCombinedChartProps(reports, data, measureIdx) {
  return data.reports.reduce(
    (prev, {id, color}) => {
      const report = reports[id];
      // skip unauthorized reports
      if (!report) {
        return prev;
      }
      let singleReportResult;
      if (data.visualization === 'number') {
        singleReportResult = [{key: report.id, value: report.result.data}];
      } else {
        singleReportResult = formatReportResult(data, report.result.measures[measureIdx].data);
      }

      return {
        resultArr: [...prev.resultArr, singleReportResult],
        reportsNames: [...prev.reportsNames, report.name],
        reportColors: [...prev.reportColors, color],
      };
    },
    {resultArr: [], reportsNames: [], reportColors: []}
  );
}
