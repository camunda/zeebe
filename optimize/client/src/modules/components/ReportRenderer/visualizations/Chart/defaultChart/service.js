/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';

const {convertToMilliseconds} = formatters;

export function getFormattedTargetValue({unit, value}) {
  if (!unit) {
    return value;
  }
  return convertToMilliseconds(value, unit);
}
