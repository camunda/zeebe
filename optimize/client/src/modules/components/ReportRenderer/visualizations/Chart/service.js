/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {getTooltipText, isDurationReport} from 'services';
import {t} from 'translation';

export function formatTooltip({
  dataset,
  dataIndex,
  configuration: {alwaysShowAbsolute, alwaysShowRelative, precision},
  formatter,
  instanceCount,
  isDuration,
  showLabel,
}) {
  if (dataset.isTarget) {
    return;
  }

  const shortNotation = isDuration ? alwaysShowAbsolute : alwaysShowAbsolute || alwaysShowRelative;

  const tooltipText = getTooltipText(
    dataset.data[dataIndex],
    formatter,
    instanceCount,
    alwaysShowAbsolute,
    alwaysShowRelative,
    isDuration,
    precision,
    shortNotation
  );

  if (!tooltipText) {
    return;
  }

  const label = showLabel && dataset.label ? dataset.label + ': ' : '';

  return shortNotation ? tooltipText : label + tooltipText;
}

export function formatTooltipTitle(title, availableWidth) {
  if (!title || !availableWidth) {
    return '';
  }

  const widthPerCharacter = 7;
  const charactersPerLine = Math.floor(availableWidth / widthPerCharacter);

  const lines = [];
  let remainingString = title;

  while (remainingString.length > charactersPerLine) {
    const currentString = remainingString.substr(0, charactersPerLine);
    let lastSpaceIndex = currentString.lastIndexOf(' ');
    const hasSuitableWhitespace = lastSpaceIndex !== -1;

    if (!hasSuitableWhitespace) {
      lastSpaceIndex = charactersPerLine;
    }

    lines.push(currentString.substr(0, lastSpaceIndex));
    remainingString = remainingString.substr(lastSpaceIndex + hasSuitableWhitespace);
  }
  lines.push(remainingString);

  return lines.join('\n');
}

export function getTooltipLabelColor({dataIndex, dataset}, type) {
  if (type === 'pie') {
    const color = dataset.backgroundColor[dataIndex];
    return {
      borderColor: color,
      backgroundColor: color,
    };
  }

  return {
    borderColor: dataset.legendColor,
    backgroundColor: dataset.legendColor,
  };
}

export function calculateLinePosition(chart) {
  const firstMeasureAxis = chart.scales['axis-0'];

  return firstMeasureAxis.getPixelForValue(chart.options.lineAt);
}

export function canBeInterpolated({type, value}, xml, decisionDefinitionKey) {
  if (type === 'flowNodes' || type === 'userTasks') {
    return false;
  }
  if (type === 'variable' && value.type === 'String') {
    return false;
  }
  if (type === 'inputVariable' || type === 'outputVariable') {
    return (
      new DOMParser()
        .parseFromString(xml, 'text/xml')
        .querySelector(
          `decision[id="${decisionDefinitionKey}"] [id="${value.id}"] ${
            type === 'inputVariable' ? 'inputExpression' : ''
          }`
        )
        .getAttribute('typeRef')
        .toLowerCase() !== 'string'
    );
  }
  return true;
}

export function getLabel({property, aggregationType, userTaskDurationTime}) {
  return (
    (userTaskDurationTime
      ? `${t('report.config.userTaskDuration.' + userTaskDurationTime)} `
      : '') +
    t('report.view.' + (property === 'frequency' ? 'count' : 'duration')) +
    (aggregationType
      ? ` - ${t('report.config.aggregationShort.' + aggregationType.type, {
          value: aggregationType.value,
        })}`
      : '')
  );
}

export function getAxisIdx(measures, measureIdx) {
  if (measures.every(({property}) => property === measures[0].property)) {
    // if every measure has the same prop, there is only one axis
    return 0;
  }
  return measures[measureIdx].property === 'frequency' ? 0 : 1;
}

export function hasReportPersistedTooltips(report) {
  const {
    data: {
      configuration: {alwaysShowAbsolute, alwaysShowRelative},
    },
  } = report;
  const isDuration = isDurationReport(report);
  return isDuration ? alwaysShowAbsolute : alwaysShowAbsolute || alwaysShowRelative;
}

export function getAxesConfig(isHorizontal) {
  return {
    axis0: {
      position: isHorizontal ? 'bottom' : 'left',
      axis: isHorizontal ? 'x' : 'y',
      id: 'axis-0',
    },
    axis1: {position: isHorizontal ? 'top' : 'right', axis: isHorizontal ? 'x' : 'y', id: 'axis-1'},
    groupBy: {
      position: isHorizontal ? 'left' : 'bottom',
      axis: isHorizontal ? 'y' : 'x',
      id: isHorizontal ? 'y' : 'x',
    },
  };
}
