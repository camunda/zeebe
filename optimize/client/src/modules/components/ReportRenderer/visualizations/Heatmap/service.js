/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';
const {convertToMilliseconds} = formatters;

export function calculateTargetValueHeat(durationData, targetValues) {
  const data = {};

  Object.keys(targetValues).forEach((element) => {
    const targetValueInMs = convertToMilliseconds(
      targetValues[element].value,
      targetValues[element].unit
    );

    if (durationData[element] > targetValueInMs) {
      data[element] = durationData[element] / targetValueInMs - 1;
    } else {
      data[element] = null;
    }
  });

  return data;
}

export function getConfig(data, flowNodeId) {
  const {
    definitions,
    configuration: {heatmapTargetValue},
  } = data;
  const {value, unit} = heatmapTargetValue.values[flowNodeId];

  return {
    processDefinitionKey: definitions[0].key,
    processDefinitionVersions: definitions[0].versions,
    tenantIds: definitions[0].tenantIds,
    includedColumns: ['processInstanceId'],
    filter: [
      ...data.filter,
      {
        type: 'flowNodeDuration',
        data: {
          [flowNodeId]: {
            operator: '>',
            value,
            unit,
          },
        },
        filterLevel: 'instance',
      },
    ],
  };
}
