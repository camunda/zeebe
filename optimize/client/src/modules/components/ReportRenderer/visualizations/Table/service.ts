/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {post} from 'request';

export async function loadObjectValues(
  name: string,
  processInstanceId: string,
  processDefinitionKey: string,
  processDefinitionVersions: string[],
  tenantIds: (string | null)[]
): Promise<string> {
  const response = await post(`api/variables/values`, {
    name,
    processInstanceId,
    processDefinitionKey,
    processDefinitionVersions,
    tenantIds,
    type: 'object',
  });

  const values = await response.json();
  return values[0];
}
