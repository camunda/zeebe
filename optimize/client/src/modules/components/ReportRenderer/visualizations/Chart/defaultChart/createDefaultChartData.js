/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters, processResult} from 'services';

import {getAxisIdx, getLabel} from '../service';
import {createDatasetOptions} from './createDefaultChartOptions';

const {formatReportResult} = formatters;

export default function createDefaultChartData(props) {
  const datasets = [];

  const {
    result: {measures = []},
    data: {
      configuration: {measureVisualizations, horizontalBar},
    },
  } = props.report;

  const chartData = measures.map((_, idx) => extractDefaultChartData(props, idx));
  const labels = chartData[0].labels;

  chartData.forEach(
    ({labels: measureLabels, formattedResult, visualization, targetValue, color, isDark}, idx) => {
      const measure = measures[idx];
      let type = visualization;
      let order;
      if (visualization === 'barLine') {
        type = measureVisualizations[measure.property];
        order = type === 'line' ? 0 : 1;
      }

      datasets.push({
        [horizontalBar ? 'xAxisID' : 'yAxisID']: 'axis-' + getAxisIdx(measures, idx),
        label: getLabel(measure),
        data: sortValues(
          labels,
          measureLabels,
          formattedResult.map(({value}) => value)
        ),
        formatter: formatters[measure.property],
        order,
        ...createDatasetOptions({
          type,
          data: formattedResult,
          targetValue,
          datasetColor: color,
          isStriped: false,
          isDark,
          measureCount: measures.length,
          datasetIdx: idx,
        }),
      });
    }
  );

  return {labels, datasets};
}

export function extractDefaultChartData({report, theme, targetValue}, measureIdx = 0) {
  const {data} = report;
  const isDark = theme === 'dark';
  const {
    visualization,
    configuration: {color},
  } = data;

  const result = processResult({...report, result: report.result.measures[measureIdx]});
  const formattedResult = formatReportResult(data, result.data);

  const labels = formattedResult.map(({key, label}) => formatters.formatLabel(label || key));

  return {
    labels,
    formattedResult,
    visualization,
    targetValue,
    color,
    isDark,
  };
}

// sorts values based on refLabels
function sortValues(refLabels, labels, values) {
  let labelValueMap = {};
  labels.forEach((label, index) => {
    labelValueMap[label] = values[index];
  });

  return refLabels.map((label) => labelValueMap[label]);
}
