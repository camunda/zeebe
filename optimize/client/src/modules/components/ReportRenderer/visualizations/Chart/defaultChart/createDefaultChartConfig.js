/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import createDefaultChartData from './createDefaultChartData';
import createDefaultChartOptions from './createDefaultChartOptions';
import createPlugins from '../createPlugins';

export default function createDefaultChartConfig(props) {
  const {
    data: {visualization},
  } = props.report;

  return {
    type: visualization === 'barLine' ? 'bar' : visualization,
    data: createDefaultChartData(props),
    options: createDefaultChartOptions(props),
    plugins: createPlugins(props),
  };
}
