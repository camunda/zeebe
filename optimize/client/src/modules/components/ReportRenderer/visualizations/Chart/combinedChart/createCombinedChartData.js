/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';
import {ColorPicker} from 'components';

import {createDatasetOptions} from '../defaultChart/createDefaultChartOptions';
import {getAxisIdx, getLabel} from '../service';
import {getCombinedChartProps} from './service';

export default function createCombinedChartData(props) {
  const {
    report: {result},
  } = props;

  if (result.measures) {
    return createMultiMeasureChartData(props);
  }

  const {labels, unitedResults, reportsNames, reportColors, targetValue, isDark, visualization} =
    extractCombinedData(props);

  const datasets = unitedResults.map((report, index) => {
    return {
      yAxisID: 'axis-0',
      label: reportsNames && reportsNames[index],
      data: report.map(({value}) => value),
      ...createDatasetOptions({
        type: visualization,
        data: report,
        targetValue,
        datasetColor: reportColors[index],
        isStriped: true,
        isDark,
      }),
    };
  });

  return {labels, datasets};
}

function createMultiMeasureChartData(props) {
  const {
    report: {
      result: {measures},
      data: {
        configuration: {measureVisualizations, stackedBar, horizontalBar},
      },
    },
  } = props;

  let labels = [];
  const datasets = [];
  const colors = ColorPicker.getGeneratedColors(measures.length * measures[0].data[0].value.length);

  measures.forEach((measure, idx) => {
    const {
      labels: measureLabels,
      unitedResults,
      reportsNames,
      targetValue,
      isDark,
      visualization,
    } = extractCombinedData(props, idx);

    let type = visualization;
    let order;
    if (visualization === 'barLine') {
      type = measureVisualizations[measure.property];
      order = type === 'line' ? 0 : 1;
    }

    unitedResults.forEach((report, index) => {
      datasets.push({
        [horizontalBar ? 'xAxisID' : 'yAxisID']: 'axis-' + getAxisIdx(measures, idx),
        label:
          reportsNames &&
          reportsNames[index] + (measures.length > 1 ? ' - ' + getLabel(measure) : ''),
        data: report.map(({value}) => value),
        formatter: formatters[measure.property],
        order,
        stack: stackedBar ? idx : undefined,
        ...createDatasetOptions({
          type,
          data: report,
          targetValue,
          datasetColor: colors[idx * unitedResults.length + index],
          isStriped: true,
          isDark,
          stackedBar,
        }),
      });
    });

    labels = measureLabels;
  });

  return {labels, datasets};
}

export function extractCombinedData({report, theme, targetValue}, measureIdx = 0) {
  const {result, data: combinedReportData} = report;

  const data = {...Object.values(result.data)[0].data, ...combinedReportData};

  const {reportsNames, resultArr, reportColors} = getCombinedChartProps(
    result.data,
    data,
    measureIdx
  );

  const isDark = theme === 'dark';
  const labelsMap = {};

  const keys = Array.from(
    new Set(
      resultArr.flat(2).map(({key, label}) => {
        labelsMap[key] = label;
        return key;
      })
    )
  );

  if (data.groupBy.type.includes('Date')) {
    keys.sort((a, b) => new Date(a) - new Date(b));
  }

  const unitedResults = uniteResults(resultArr, keys);
  let labels =
    data.visualization === 'number' ? reportsNames : keys.map((key) => labelsMap[key] || key);
  labels = labels.map((label) => formatters.formatLabel(label));

  return {
    labels,
    unitedResults,
    reportsNames,
    reportColors,
    targetValue,
    isDark,
    visualization: data.visualization,
  };
}

function uniteResults(results, allKeys) {
  const unitedResults = [];
  results.forEach((result) => {
    const resultObj = formatters.objectifyResult(result);
    const newResult = [];
    allKeys.forEach((key) => {
      if (typeof resultObj[key] === 'undefined') {
        newResult.push({key, value: null});
      } else {
        newResult.push({key, value: resultObj[key]});
      }
    });
    unitedResults.push(newResult);
  });

  return unitedResults;
}
