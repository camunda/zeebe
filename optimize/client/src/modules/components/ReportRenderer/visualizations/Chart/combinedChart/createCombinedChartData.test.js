/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {createDatasetOptions} from '../defaultChart/createDefaultChartOptions';
import {getCombinedChartProps} from './service';
import createCombinedChartData from './createCombinedChartData';

jest.mock('../defaultChart/createDefaultChartOptions', () => ({createDatasetOptions: jest.fn()}));
jest.mock('./service', () => ({getCombinedChartProps: jest.fn()}));

const createReport = ({reportA, reportB, groupByType}, measures, visualization = 'line') => {
  getCombinedChartProps.mockReturnValue({
    reportsNames: ['report A', 'report B'],
    resultArr: [reportA.data, reportB.data],
    reportColors: [],
  });

  return {
    result: {
      data: {
        reportA: {
          name: 'Report A',
          result: {
            data: reportA.data,
          },
        },
        reportB: {
          name: 'Report B',
          result: {
            data: reportB.data,
          },
        },
      },
      measures,
    },
    data: {
      groupBy: {
        type: groupByType || 'startDate',
        value: 'month',
      },
      reports: [
        {id: 'reportA', color: reportA.color},
        {id: 'reportB', color: reportB.color},
      ],
      visualization,
      configuration: {
        measureVisualizations: {frequency: 'line', duration: 'bar'},
        stackedBar: true,
      },
    },
    combined: true,
  };
};

it('should return correct chart data object for a combined report', () => {
  const reportA = {
    data: [
      {key: 'foo', value: 123, label: 'Flownode Foo'},
      {key: 'bar', value: 5, label: 'Barrrr'},
    ],
    color: 'blue',
  };

  const reportB = {
    data: [
      {key: 'foo', value: 1, label: 'Flownode Foo'},
      {key: 'dar', value: 3, label: 'Flownode DAR'},
    ],
    color: 'yellow',
  };

  const chartData = createCombinedChartData({
    report: createReport({reportA, reportB, groupByType: 'flowNodes'}),
    targetValue: false,
    theme: 'light',
  });

  expect(chartData).toMatchSnapshot();
});

it('should return correct chart data object for a combined report with correct date order', () => {
  const reportA = {
    data: [{key: '2017-12-27T14:21:56.000', value: 123, label: 'Dec 2017'}],
    color: 'blue',
  };

  const reportB = {
    data: [{key: '2017-03-27T14:21:56.000', value: 5, label: 'Mar 2017'}],
    color: 'yellow',
  };

  const chartData = createCombinedChartData({
    report: createReport({reportA, reportB}),
    targetValue: false,
    theme: 'light',
  });

  expect(chartData).toMatchSnapshot();
});

it('should return a dataset for each measure value in multi measure reports', () => {
  const reportA = {
    data: [{key: 'flowNode1', value: 123, label: 'Dec 2017'}],
    color: 'blue',
  };

  const reportB = {
    data: [{key: 'flowNode2', value: 5, label: 'Mar 2017'}],
    color: 'yellow',
  };

  const measures = [
    {
      property: 'frequency',
      data: [{key: 'flowNode1', value: [{key: 'assignee1', value: 123}]}],
    },
    {
      property: 'duration',
      data: [{key: 'flowNode1', value: [{key: 'assignee1', value: 5536205036}]}],
    },
  ];

  const chartData = createCombinedChartData({
    report: createReport({reportA, reportB}, measures),
    targetValue: false,
    theme: 'light',
  });

  expect(chartData.datasets.length).toBe(4);
  expect(chartData.datasets[0]).toEqual({
    yAxisID: 'axis-0',
    label: 'report A - Count',
    data: [123, null],
    formatter: expect.any(Function),
    order: undefined,
    stack: 0,
  });
  expect(chartData.datasets[2]).toEqual({
    yAxisID: 'axis-1',
    label: 'report A - Duration',
    data: [123, null],
    formatter: expect.any(Function),
    order: undefined,
    stack: 1,
  });
});

it('should assign line/bar visualization to dataset according to measureVisualizations configuration', () => {
  createDatasetOptions.mockImplementation((options) => options);
  const reportA = {
    data: [{key: 'flowNode1', value: 123, label: 'Dec 2017'}],
    color: 'blue',
  };

  const reportB = {
    data: [{key: 'flowNode2', value: 5, label: 'Mar 2017'}],
    color: 'yellow',
  };

  const measures = [
    {
      property: 'frequency',
      data: [{key: 'flowNode1', value: [{key: 'assignee1', value: 123}]}],
    },
    {
      property: 'duration',
      data: [{key: 'flowNode1', value: [{key: 'assignee1', value: 5536205036}]}],
    },
  ];

  const chartData = createCombinedChartData({
    report: createReport({reportA, reportB}, measures, 'barLine'),
    targetValue: false,
    theme: 'light',
  });

  const getDataset = (label) => chartData.datasets.find((dataset) => dataset.label === label);
  expect(getDataset('report A - Count').type).toBe('line');
  expect(getDataset('report A - Count').order).toBe(0);
  expect(getDataset('report B - Count').type).toBe('line');
  expect(getDataset('report B - Count').order).toBe(0);
  expect(getDataset('report A - Duration').type).toBe('bar');
  expect(getDataset('report A - Duration').order).toBe(1);
  expect(getDataset('report B - Duration').type).toBe('bar');
  expect(getDataset('report B - Duration').order).toBe(1);
});
