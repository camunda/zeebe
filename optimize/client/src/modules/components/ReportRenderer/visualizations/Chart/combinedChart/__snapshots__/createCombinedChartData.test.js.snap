// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct chart data object for a combined report 1`] = `
Object {
  "datasets": Array [
    Object {
      "data": Array [
        123,
        5,
        null,
      ],
      "label": "report A",
      "yAxisID": "axis-0",
    },
    Object {
      "data": Array [
        1,
        null,
        3,
      ],
      "label": "report B",
      "yAxisID": "axis-0",
    },
  ],
  "labels": Array [
    "Flownode Foo",
    "Barrrr",
    "Flownode DAR",
  ],
}
`;

exports[`should return correct chart data object for a combined report with correct date order 1`] = `
Object {
  "datasets": Array [
    Object {
      "data": Array [
        null,
        123,
      ],
      "label": "report A",
      "yAxisID": "axis-0",
    },
    Object {
      "data": Array [
        5,
        null,
      ],
      "label": "report B",
      "yAxisID": "axis-0",
    },
  ],
  "labels": Array [
    "Mar 2017",
    "Dec 2017",
  ],
}
`;
